tests/wpt/metadata/html/semantics/forms/form-submission-0/constructing-form-data-set.html.ini

```diff
@@ -10,13 +10,3 @@
 
   [Entries added to "formData" IDL attribute should be submitted.]
```

Comment:
This test case still gets failure.

https://github.com/servo/servo/blob/4ac5b8a3aea5df3bf374dcfdb22c34f5ff0bafe8/tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html#L98-L109

I tried to add logs to check if the event is dispatched and it's triggered correctly.
So I'm wondering if the `search` is not updated correctly after submitting the form.

Reply:
Ok. I've found what exact the wrong output is, will keep investigating the root cause.

The expected search is `n1=v1&h1=vh1` but we got `h1=vh1&n1=v1`.

Reply:
I've observed one issue here but the test case still got failure.

About the observed issue is, we're using `BTreeMap` for `FormData` so when we do iteration for it, the data will be sorted automatically. I'm wondering maybe we need to change to `IndexMap` instead?

https://github.com/servo/servo/blob/4d94692b043479ef420c51eea2b3fc9e7e709232/components/script/dom/formdata.rs#L28

After observed above issue, I tried to use `IndexMap` but I found the test case still got failure. Then I tried to make a break point in this line

https://github.com/servo/servo/blob/4d94692b043479ef420c51eea2b3fc9e7e709232/components/script/dom/htmlformelement.rs#L471

```
(lldb) p *load_data.url.__0.ptr.pointer
(alloc::sync::ArcInner<url::Url>) $1 = {
  strong = {
    v = (value = 1)
  }
  weak = {
    v = (value = 1)
  }
  data = {
    serialization = {
      vec = {
        buf = {
          ptr = (pointer = "http://web-platform.test:8000/common/blank.html?n1=v1&h1=vh1", _marker = core::marker::PhantomData<unsigned char> @ 0x00007fbc259b51f0)
          cap = 94
        }
        len = 60
      }
    }
    scheme_end = 4
    username_end = 7
    host_start = 7
    host_end = 24
    host = {}
    port = {}
    path_start = 29
    query_start = {}
    fragment_start = {}
  }
}
```

I found the url was shown correctly with `?n1=v1&h1=vh1`.

But, when I added a log `console.log(iframe.contentWindow.location.href);` in the `iframe.onload` event listener. It shows the `href` is `about:blank` so the `location.search` would be empty string which will not be equal to `n1=v1&h1=vh1`.

I'm wondering if the onload event is triggered after the form submitted 🤔

Reply:
Ok, I did some investigation around iframe load events (which caused me to read through https://github.com/whatwg/html/issues/490 and remember a bunch of awful details). As far as I can tell, the reason we observe the about:blank behaviour in this test is that there is a load event queued for the iframe immediately when it is added to the document (for the default about:blank browsing context). I created a test to verify this theory:
```html
<script src="/resources/testharness.js"></script>
<script src="/resources/testharnessreport.js"></script>
<iframe name="i" id="i"></iframe>
<form id="f" action="/common/blank.html" target="i"></form>
<script>
  async_test(function(t) {
    var i = document.getElementById('i');
    i.onload = t.step_func(function() {
      if (i.contentWindow.location.href == "about:blank") { return; } 
      assert_equals(i.contentWindow.location.pathname, "/common/blank.html"); 
      t.done();
    });
    document.getElementById('f').submit();
  });
</script>
```
This passes in Servo for me today. The spec says that the iframe load event should be queued when it's added to the document, and that form submission causes a task to be queued (and that task initiates a navigation for the iframe's browsing context), so it seems perfectly valid for the about:blank's load event to fire during that time before the iframe has officially been navigated away from about:blank.

THEREFORE

I think the right solution here is to add a comment with a link to https://github.com/whatwg/html/issues/490 and add a check for about:blank in the iframe load event handler and ignore the event if it matches.

Reply:
@jdm thanks a lot for helping me investigate this! That's clear for me now that why the test failed; I think we can only update the test and remove the comment once the issue is fixed.